sdk/core/core/src/az_http_request_builder.c

```diff
@@ -12,84 +12,231 @@
+                                      .headers_end = 0 };
+
   return AZ_OK;
 }
 
-/**
- * @brief
- *
- * @param p_hrb
- * @param name
- * @param value
- * @return az_result
- */
 az_result az_http_request_builder_set_query_parameter(
     az_http_request_builder * const p_hrb,
     az_const_span const name,
     az_const_span const value) {
-  (void)p_hrb;
-  (void)name;
-  (void)value;
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  if (!az_const_span_is_valid(name) || !az_const_span_is_valid(value) || name.size == 0) {
+    return AZ_ERROR_ARG;
+  }
+
+  size_t new_url_size;
+  bool const has_value = value.size > 0;
+  {
+    az_span new_url_span = { .begin = p_hrb->url.begin };
+    size_t const extra_chars_size = sizeof((uint8_t)'?') + (has_value ? sizeof((uint8_t)'=') : 0);
+    size_t const name_and_value_size = name.size + value.size;
+    size_t const appended_size = name_and_value_size + extra_chars_size;
+
+    new_url_size = p_hrb->url.size + appended_size;
+
+    if (name_and_value_size < name.size || name_and_value_size < value.size
+        || appended_size < name_and_value_size || appended_size < extra_chars_size
+        || new_url_size < appended_size || new_url_size < p_hrb->url.size
+        || new_url_size > p_hrb->max_url_size) {
+      return AZ_ERROR_BUFFER_OVERFLOW;
+    }
+
+    new_url_span.size = new_url_size;
+
+    if (az_const_span_is_overlap(az_span_to_const_span(new_url_span), name)
+        || az_const_span_is_overlap(az_span_to_const_span(new_url_span), value)) {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  bool first_parameter = true;
+  for (size_t i = p_hrb->url.size; i > 0;) {
+    --i;
+    if (p_hrb->url.begin[i] == '?') {
+      first_parameter = false;
+    }
+  }
+
+  p_hrb->url.begin[p_hrb->url.size] = first_parameter ? '?' : '&';
+  ++(p_hrb->url.size);
+
+  memcpy(p_hrb->url.begin + p_hrb->url.size, name.begin, name.size);
+  p_hrb->url.size += name.size;
+
+  if (has_value) {
+    p_hrb->url.begin[p_hrb->url.size] = '=';
+    ++(p_hrb->url.size);
+
+    memcpy(p_hrb->url.begin + p_hrb->url.size, value.begin, value.size);
+    p_hrb->url.size += value.size;
+  }
+
+  assert(p_hrb->url.size == new_url_size);
   return AZ_OK;
 }
 
-/**
- * @brief
- *
- * @param p_hrb
- * @param name
- * @param value
- * @return az_result
- */
 az_result az_http_request_builder_append_header(
     az_http_request_builder * const p_hrb,
     az_const_span const name,
     az_const_span const value) {
-  (void)p_hrb;
-  (void)name;
-  (void)value;
+  AZ_CONTRACT_ARG_NOT_NULL(p_hrb);
+
+  if (p_hrb->headers_end >= p_hrb->max_headers) {
+    return AZ_ERROR_BUFFER_OVERFLOW;
+  }
+
+  if (!az_const_span_is_valid(name) || !az_const_span_is_valid(value) || name.size == 0
+      || value.size == 0) {
+    return AZ_ERROR_ARG;
+  }
+
+  for (size_t i = 0; i < name.size; ++i) {
+    if (name.begin[i] == '\0') {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  for (size_t i = 0; i < value.size; ++i) {
```

Comment:
IMHO, we should create functions instead of copy-paste code, something like `az_span_find()`, `az_span_is_str()`?

Reply:
I don't see a reason to check for any \0's. Why have these loops at all?

Reply:
This no longer applies now that we store pairs in the buffer instead of copying values.